packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,246 @@
-import { sendRequest, HttpMethod } from '@biconomy/common'
+import { Logger, sendRequest, HttpMethod, getTimestampInSeconds } from '@biconomy/common'
 import { resolveProperties } from '@ethersproject/properties'
-import { UserOperation } from '@biconomy/core-types'
-import { IPaymaster } from 'interfaces/IPaymaster'
-import { PaymasterAndDataResponse, PaymasterConfig } from './utils/Types'
+import { UserOperation, Transaction } from '@biconomy/core-types'
+import { Provider } from '@ethersproject/abstract-provider'
+import {
+  PaymasterFeeQuote,
+  PaymasterConfig,
+  BiconomyTokenPaymasterFeeQuoteResponse,
+  FeeQuotesOrDataDto,
+  SponsorUserOperationDto,
+  PaymasterServiceSuccessResponse,
+  BiconomyTokenPaymasterRequest,
+  PaymasterMode
+} from './utils/Types'
+import { BigNumberish, BigNumber, ethers } from 'ethers'
+import { ERC20_ABI } from './constants'
+import { IHybridPaymaster } from './interfaces/IHybridPaymaster'
 
-export class BiconomyPaymaster implements IPaymaster {
+// Hybrid - Generic Gas Abstraction paymaster
+/**
+ *
+ */
+export class BiconomyPaymaster implements IHybridPaymaster {
   constructor(readonly paymasterConfig: PaymasterConfig) {}
+
+  // Note: maybe rename to createTokenApprovalTransaction
+  async createTokenApprovalRequest(
+    tokenPaymasterRequest: BiconomyTokenPaymasterRequest,
+    provider: Provider
+  ): Promise<Transaction> {
+    const feeTokenAddress: string = tokenPaymasterRequest.feeQuote.tokenAddress
+    Logger.log('erc20 fee token address ', feeTokenAddress)
+
+    const spender = tokenPaymasterRequest.spender
+    Logger.log('spender address ', spender)
+
+    // TODO
+    // Note: should also check in caller if the approval is already given, if yes return object with address or data 0
+    // Note: we would need userOp here to get the account/owner info to check allowance
+
+    let requiredApproval: BigNumberish = BigNumber.from(0)
+
+    if (tokenPaymasterRequest.maxApproval && tokenPaymasterRequest.maxApproval == true) {
+      requiredApproval = ethers.constants.MaxUint256
+    } else {
+      /*requiredApproval = ethers.BigNumber.from(
+        tokenPaymasterRequest.feeQuote.maxGasFee.toString()
+      ).mul(
+        ethers.BigNumber.from(10).pow(
+          ethers.BigNumber.from(tokenPaymasterRequest.feeQuote.decimal.toString())
+        )
+      )*/
+
+      requiredApproval = BigNumber.from(
+        Math.floor(
+          tokenPaymasterRequest.feeQuote.maxGasFee *
+            Math.pow(10, tokenPaymasterRequest.feeQuote.decimal)
+        )
+      )
+    }
+
+    Logger.log('required approval for erc20 token ', requiredApproval)
+
+    const erc20 = new ethers.Contract(feeTokenAddress, ERC20_ABI, provider)
```

Comment:
No need to create a token instance here. You can use just the interface object only like this

```
// Create an Interface instance
const iface = new Interface(ERC20_ABI);

// Encode the function data
const data = iface.encodeFunctionData('approve', [spender, requiredApproval]);

```

Reply:
right I don't need a provider. for allowance check I would need to read so I had kept like this